# <font color='blue'>Data Science Academy - Python Fundamentos - Capítulo 8</font>

## Download: http://github.com/dsacademybr

### NumPy

Para importar numpy, utilize: 
import numpy as np


Você também pode utilizar: 
from numpy import *  . Isso evitará a utilização de np., mas este comando importará todos os módulos do NumPy.

Para atualizar o NumPy, abra o prompt de comando e digite: pip install numpy -U

In [1]:
# Importando o NumPy
import numpy as np

In [2]:
np.__version__

'1.20.1'

## Criando Arrays

In [3]:
# Help
help(np.array)

Help on built-in function array in module numpy:

array(...)
    array(object, dtype=None, *, copy=True, order='K', subok=False, ndmin=0,
          like=None)
    
    Create an array.
    
    Parameters
    ----------
    object : array_like
        An array, any object exposing the array interface, an object whose
        __array__ method returns an array, or any (nested) sequence.
    dtype : data-type, optional
        The desired data-type for the array.  If not given, then the type will
        be determined as the minimum type required to hold the objects in the
        sequence.
    copy : bool, optional
        If true (default), then the object is copied.  Otherwise, a copy will
        only be made if __array__ returns a copy, if obj is a nested sequence,
        or if a copy is needed to satisfy any of the other requirements
        (`dtype`, `order`, etc.).
    order : {'K', 'A', 'C', 'F'}, optional
        Specify the memory layout of the array. If object is not an array

In [4]:
# Array criado a partir de uma lista:
# Um array é um conjunto de valores do mesmo tipo indexados por uma tupla de valores. São semelhante a listas em python, mas não são tão flexiveis. Todos os elementos devem ser do mesmo tipo numerico (int, ou float)
# Possuem alto desempenho para grandes conjuntos de dados numericos
# Transformar uma lista em array nos da muitos metodos interessantes para trabalhar com os valores
vetor1 = np.array([0, 1, 2, 3, 4, 5, 6, 7, 8])

In [5]:
# Array unidemensional
print(vetor1)

[0 1 2 3 4 5 6 7 8]


In [6]:
# Um objeto do tipo ndarray é um recipiente multidimensional de itens do mesmo tipo e tamanho.
type(vetor1)

numpy.ndarray

In [7]:
# Usando métodos do array NumPy. Este metodo soma cada elemento com os anteriores para o numero de elementos totais
# Soma acumulada
vetor1.cumsum()

array([ 0,  1,  3,  6, 10, 15, 21, 28, 36], dtype=int32)

In [8]:
# Criando uma lista. Perceba como listas e arrays são objetos diferentes, com diferentes propriedades
lst = [0, 1, 2, 3, 4, 5, 6, 7, 8]

In [9]:
lst

[0, 1, 2, 3, 4, 5, 6, 7, 8]

In [13]:
type(lst)

list

In [14]:
# Imprimindo na tela um elemento específico no array
vetor1[0] 

0

In [15]:
# Alterando um elemento do array
vetor1[0] = 100

In [16]:
print(vetor1)

[100   1   2   3   4   5   6   7   8]


In [17]:
# Não é possível incluir elemento de outro tipo
vetor1[0] = 'Novo elemento'

ValueError: invalid literal for int() with base 10: 'Novo elemento'

In [18]:
# Verificando o formato do array
# Como é unidimensional vai aparecer apenas o numero de elementos da primeira dimensão
print(vetor1.shape)

(9,)


## Funções NumPy

In [19]:
# A função arange cria um vetor contendo uma progressão aritmética a partir de um intervalo - start, stop, step
# é semelhante a função pyhton range(), função built-in
vetor2 = np.arange(0., 4.5, .5)

In [20]:
print(vetor2)

[0.  0.5 1.  1.5 2.  2.5 3.  3.5 4. ]


In [21]:
# Verificando o tipo do objeto
type(vetor2)

numpy.ndarray

In [22]:
# Formato do array
np.shape(vetor2)

(9,)

In [23]:
print (vetor2.dtype)

float64


In [24]:
x = np.arange(1, 10, 0.25)
print(x)

[1.   1.25 1.5  1.75 2.   2.25 2.5  2.75 3.   3.25 3.5  3.75 4.   4.25
 4.5  4.75 5.   5.25 5.5  5.75 6.   6.25 6.5  6.75 7.   7.25 7.5  7.75
 8.   8.25 8.5  8.75 9.   9.25 9.5  9.75]


In [25]:
# Array preenchido com zeros
print(np.zeros(10))

[0. 0. 0. 0. 0. 0. 0. 0. 0. 0.]


In [34]:
# Retorna 1 nas posições em diagonal e 0 no restante
# O (3) modifica a quantidade de linhas formando uma matriz bidimensional se for acima de 1, pode ser modificado
z = np.eye(3)

In [35]:
z

array([[1., 0., 0.],
       [0., 1., 0.],
       [0., 0., 1.]])

In [36]:
# Os valores passados como parâmetro, formam uma diagonal usando np.diag() onde os parametros devem ser meu array
d = np.diag(np.array([1, 2, 3, 4]))

In [37]:
d

array([[1, 0, 0, 0],
       [0, 2, 0, 0],
       [0, 0, 3, 0],
       [0, 0, 0, 4]])

In [38]:
# Array de números complexos
c = np.array([1+2j, 3+4j, 5+6*1j])

In [39]:
c

array([1.+2.j, 3.+4.j, 5.+6.j])

In [40]:
# Array de valores booleanos
b = np.array([True, False, False, True])

In [41]:
b

array([ True, False, False,  True])

In [44]:
# Array de strings e usando o np.diag()
s = np.diag(np.array(['Python', 'R', 'Julia']))

In [45]:
s

array([['Python', '', ''],
       ['', 'R', ''],
       ['', '', 'Julia']], dtype='<U6')

In [48]:
# O método linspace (linearly spaced vector) retorna um número de 
# valores igualmente distribuídos no intervalo especificado 
np.linspace(0, 10)

array([ 0.        ,  0.20408163,  0.40816327,  0.6122449 ,  0.81632653,
        1.02040816,  1.2244898 ,  1.42857143,  1.63265306,  1.83673469,
        2.04081633,  2.24489796,  2.44897959,  2.65306122,  2.85714286,
        3.06122449,  3.26530612,  3.46938776,  3.67346939,  3.87755102,
        4.08163265,  4.28571429,  4.48979592,  4.69387755,  4.89795918,
        5.10204082,  5.30612245,  5.51020408,  5.71428571,  5.91836735,
        6.12244898,  6.32653061,  6.53061224,  6.73469388,  6.93877551,
        7.14285714,  7.34693878,  7.55102041,  7.75510204,  7.95918367,
        8.16326531,  8.36734694,  8.57142857,  8.7755102 ,  8.97959184,
        9.18367347,  9.3877551 ,  9.59183673,  9.79591837, 10.        ])

In [55]:
# Alterando os parametros, (start, stop, quantidade de valores)
print(np.linspace(0, 10, 8))

[ 0.          1.42857143  2.85714286  4.28571429  5.71428571  7.14285714
  8.57142857 10.        ]


In [56]:
# Colocando os valores em Log
print(np.logspace(0, 10, 8))

[1.00000000e+00 2.68269580e+01 7.19685673e+02 1.93069773e+04
 5.17947468e+05 1.38949549e+07 3.72759372e+08 1.00000000e+10]


## Criando Matrizes

In [57]:
# Criando uma matriz
# Uma matriz é um array bidimensional
matriz = np.array([[1,2,3],[4,5,6]]) 

In [58]:
print(matriz)

[[1 2 3]
 [4 5 6]]


In [59]:
# Shape é atributo, não é metodo, por isso não tem (). Matriz 2 por 3
print(matriz.shape)

(2, 3)


In [60]:
# Criando uma matriz 2x3 apenas com números "1"
matriz1 = np.ones((2,3))

In [61]:
print(matriz1)

[[1. 1. 1.]
 [1. 1. 1.]]


In [62]:
matrizz = np.zeros((2,3))

In [63]:
print(matrizz)

[[0. 0. 0.]
 [0. 0. 0.]]


In [64]:
# Criando uma matriz a partir de uma lista de listas
lista = [[13,81,22], [0, 34, 59], [21, 48, 94]]

In [65]:
# A função matrix cria uma matriz a partir de uma sequência
matriz2 = np.matrix(lista)

In [66]:
matriz2

matrix([[13, 81, 22],
        [ 0, 34, 59],
        [21, 48, 94]])

In [67]:
type(matriz2)

numpy.matrix

In [68]:
# Formato da matriz. Aqui o Shpe é um método, usando como parametro a variavel d einteresse
np.shape(matriz2)

(3, 3)

In [69]:
matriz2.size

9

In [70]:
print(matriz2.dtype)

int32


In [71]:
# Tamanho que será ocupado na memoria do pc
matriz2.itemsize

4

In [72]:
matriz2.nbytes

36

In [73]:
# Um elemento da matriz (slicing), coluna 2 indice 1
print(matriz2[2,1])

48


In [74]:
# Alterando um elemento da matriz
matriz2[1,0] = 100

In [75]:
matriz2

matrix([[ 13,  81,  22],
        [100,  34,  59],
        [ 21,  48,  94]])

In [77]:
x = np.array([1, 2])  # NumPy decide o tipo dos dados
y = np.array([1.0, 2.0])  # NumPy decide o tipo dos dados
z = np.array([1, 2], dtype=np.float64)  # Forçamos um tipo de dado em particular

print (x.dtype, y.dtype, z.dtype)

int32 float64 float64


In [78]:
matriz3 = np.array([[24, 76], [35, 89]], dtype=float)

In [79]:
matriz3

array([[24., 76.],
       [35., 89.]])

In [80]:
matriz3.itemsize

8

In [81]:
matriz3.nbytes

32

In [82]:
#Total de dimensões
matriz3.ndim

2

In [83]:
matriz3[1,1]

89.0

In [84]:
matriz3[1,1] = 100

In [85]:
matriz3

array([[ 24.,  76.],
       [ 35., 100.]])

## Usando o Método random() do NumPy

In [ ]:
print(np.random.rand(10))

In [ ]:
import matplotlib.pyplot as plt
%matplotlib inline

In [ ]:
print(np.random.rand(10))

In [ ]:
plt.show((plt.hist(np.random.rand(1000))))

In [ ]:
print(np.random.randn(5,5))

In [ ]:
plt.show(plt.hist(np.random.randn(1000)))

In [ ]:
imagem = np.random.rand(30, 30)
plt.imshow(imagem, cmap = plt.cm.hot)    
plt.colorbar()    

## Operações com datasets

In [ ]:
import os
filename = os.path.join('iris.csv')

In [ ]:
# No Windows use !more iris.csv. Mac ou Linux use !head iris.csv
!head iris.csv
#!more iris.csv

In [ ]:
# Carregando um dataset para dentro de um array
arquivo = np.loadtxt(filename, delimiter=',', usecols=(0,1,2,3), skiprows=1)
print (arquivo)

In [ ]:
type(arquivo)

In [ ]:
# Gerando um plot a partir de um arquivo usando o NumPy
var1, var2 = np.loadtxt(filename, delimiter=',', usecols=(0,1), skiprows=1, unpack=True)
plt.show(plt.plot(var1, var2, 'o', markersize=8, alpha=0.75))

## Estatística

In [ ]:
# Criando um array
A = np.array([15, 23, 63, 94, 75])

In [ ]:
# Em estatística a média é o valor que aponta para onde mais se concentram os dados de uma distribuição.
np.mean(A)

In [ ]:
# O desvio padrão mostra o quanto de variação ou "dispersão" existe em 
# relação à média (ou valor esperado). 
# Um baixo desvio padrão indica que os dados tendem a estar próximos da média.
# Um desvio padrão alto indica que os dados estão espalhados por uma gama de valores.
np.std(A)

In [ ]:
# Variância de uma variável aleatória é uma medida da sua dispersão 
# estatística, indicando "o quão longe" em geral os seus valores se 
# encontram do valor esperado
np.var(A)

In [ ]:
d = np.arange(1, 10)

In [ ]:
d

In [ ]:
np.sum(d)

In [ ]:
# Retorna o produto dos elementos
np.prod(d)

In [ ]:
# Soma acumulada dos elementos
np.cumsum(d)

In [ ]:
a = np.random.randn(400,2)
m = a.mean(0)
print (m, m.shape)

In [ ]:
plt.plot(a[:,0], a[:,1], 'o', markersize=5, alpha=0.50)
plt.plot(m[0], m[1], 'ro', markersize=10)
plt.show()

## Outras Operações com Arrays

In [ ]:
# Slicing
a = np.diag(np.arange(3))

In [ ]:
a

In [ ]:
a[1, 1]

In [ ]:
a[1]

In [ ]:
b = np.arange(10)

In [ ]:
b

In [ ]:
# [start:end:step]
b[2:9:3] 

In [ ]:
# Comparação
a = np.array([1, 2, 3, 4])
b = np.array([4, 2, 2, 4])
a == b

In [ ]:
np.array_equal(a, b)

In [ ]:
a.min()

In [ ]:
a.max()

In [ ]:
# Somando um elemento ao array
np.array([1, 2, 3]) + 1.5

In [ ]:
# Usando o método around
a = np.array([1.2, 1.5, 1.6, 2.5, 3.5, 4.5])

In [ ]:
b = np.around(a)

In [ ]:
b

In [ ]:
# Criando um array
B = np.array([1, 2, 3, 4])

In [ ]:
B

In [ ]:
# Copiando um array
C = B.flatten()

In [ ]:
C

In [ ]:
# Criando um array
v = np.array([1, 2, 3])

In [ ]:
# Adcionando uma dimensão ao array
v[:, np.newaxis], v[:,np.newaxis].shape, v[np.newaxis,:].shape

In [ ]:
# Repetindo os elementos de um array
np.repeat(v, 3)

In [ ]:
# Repetindo os elementos de um array
np.tile(v, 3)

In [ ]:
# Criando um array
w = np.array([5, 6])

In [ ]:
# Concatenando
np.concatenate((v, w), axis=0)

In [ ]:
# Copiando arrays
r = np.copy(v)

In [ ]:
r

Conheça a Formação Cientista de Dados, um programa completo, 100% online e 100% em português, com 340 horas, mais de 1.200 aulas em vídeos e 26 projetos, que vão ajudá-lo a se tornar um dos profissionais mais cobiçados do mercado de análise de dados. Clique no link abaixo, faça sua inscrição, comece hoje mesmo e aumente sua empregabilidade:

https://www.datascienceacademy.com.br/pages/formacao-cientista-de-dados

# Fim

### Obrigado - Data Science Academy - <a href="http://facebook.com/dsacademybr">facebook.com/dsacademybr</a>